contracts/examples/mintable-erc1155-transfer/FxMintableERC1155RootTunnel.sol

```diff
@@ -0,0 +1,172 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {ERC1155} from "../../lib/ERC1155.sol";
+import {ERC1155Holder} from "../../lib/ERC1155Holder.sol";
+import {Create2} from "../../lib/Create2.sol";
+import {FxBaseRootTunnel} from "../../tunnel/FxBaseRootTunnel.sol";
+import {Address} from "../../lib/Address.sol";
+
+contract FxMintableERC1155RootTunnel is FxBaseRootTunnel, Create2, ERC1155Holder {
+    bytes32 public constant DEPOSIT = keccak256("DEPOSIT");
+    bytes32 public constant DEPOSIT_BATCH = keccak256("DEPOSIT_BATCH");
+    bytes32 public constant WITHDRAW = keccak256("WITHDRAW");
+    bytes32 public constant WITHDRAW_BATCH = keccak256("WITHDRAW_BATCH");
+    bytes32 public constant MAP_TOKEN = keccak256("MAP_TOKEN");
+
+    event TokenMappedMintableERC1155(address indexed rootToken, address indexed childToken);
+    event FxWithdrawMintableERC1155(
+        address indexed rootToken,
+        address indexed childToken,
+        address indexed userAddress,
+        uint256 id,
+        uint256 amount
+    );
+    event FxDepositMintableERC1155(
+        address indexed rootToken,
+        address indexed depositor,
+        address indexed userAddress,
+        uint256 id,
+        uint256 amount
+    );
+    event FxWithdrawBatchMintableERC1155(
+        address indexed rootToken,
+        address indexed childToken,
+        address indexed userAddress,
+        uint256[] ids,
+        uint256[] amounts
+    );
+    event FxDepositBatchMintableERC1155(
+        address indexed rootToken,
+        address indexed userAddress,
+        uint256[] ids,
+        uint256[] amounts
+    );
+
+    mapping(address => address) public rootToChildTokens;
+    bytes32 public immutable childTokenTemplateCodeHash;
+
+    constructor(
+        address _checkpointManager,
+        address _fxRoot,
+        address _fxERC1155Token
+    ) FxBaseRootTunnel(_checkpointManager, _fxRoot) {
+        childTokenTemplateCodeHash = keccak256(minimalProxyCreationCode(_fxERC1155Token));
+    }
+
+    function mapToken(address rootToken) public {
```

Comment:
Consider this griefing scenario -:

A user deploys a root token with URI U1 through `mapToken` function, but griefer notices this and uses `FxMintableERC721ChildTunnel.deployChildToken` to deploy a child token with URI U2. Now across both chains, URI of the connected token is different which it shouldn't be.

I would suggest to make `mapToken` internal to avoid this similar to `FxMintableERC721` where the deployment of Root Token should be handled internally.

Reply:
Yep, you're right. bd287b9ea746c1b69d7a08d174a10003d15bc70c
Handling map token like `FxMintableERC721`, where rootToken will be deployed by tunnel contract on receiving WITHDRAW/BATCH_WITHDRAW messages. cc @QEDK